articles/142_idl.md

```diff
@@ -144,3 +144,95 @@ For now only a one-dimensional, fixed-size array is supported though.
+
+| IDL type    | C type           | C++ type                       | Python type          |
+| --------    | -----------      | ------------------------------ | -------------------- |
+| float       | float            | float                          | float                |
+| double      | double           | double                         | float                |
+| long double | long double      | long double[<sup>2</sup>](#ld) | float                |
+| char        | unsigned char    | unsigned char                  | str with length 1    |
+| wchar       | char16\_t        | char16\_t                      | str with length 1    |
+| bool        | \_Bool           | bool                           | bool                 |
+| octet       | unsigned char    | std::byte[<sup>1</sup>](#byte) | bytes with length 1  |
+| int8        | int8\_t          | int8\_t                        | int                  |
+| uint8       | uint8\_t         | uint8\_t                       | int                  |
+| int16       | int16\_t         | int16\_t                       | int                  |
+| uint16      | uint16\_t        | uint16\_t                      | int                  |
+| int32       | int32\_t         | int32\_t                       | int                  |
+| uint32      | uint32\_t        | uint32\_t                      | int                  |
+| int64       | int64\_t         | int64\_t                       | int                  |
+| uint64      | uint64\_t        | uint64\_t                      | int                  |
+
+1. <a name="byte"></a>If std::byte is not available then unsigned char is used instead.
+2. <a name="ld"></a>Users should research the support for `long double` with their choice of middleware and platform.
+   For example, it is only [64 bits when using Visual Studio](https://docs.microsoft.com/en-us/cpp/c-language/type-long-double?view=vs-2017).
+
+The next table defines how IDL templated and constructed types are mapped to the programming languages (unless specified otherwise below).
+When specified `T` is either one of the types above or an IDL struct.
+`N` is the upper limit of a bounded type.
+
+| IDL type       | C type                            | C++ type         | Python type |
+| -------------- | --------------------------------- | ---------------- | ----------- |
+| T[N]           | T[N]                              | std::array<T, N> | list        |
+| sequence<T>    | struct {size\_t, T * }            | std::vector<T>   | list        |
+| sequence<T, N> | struct {size\_t, T * }, size\_t N | std::vector<T>   | list        |
+| string         | char *                            | std::string      | str         |
+| string<N>      | char *                            | std::string      | str         |
+| wstring        | char16\_t *                       | std::u16string   | str         |
+| wstring<N>     | char16\_t *                       | std::u16string   | str         |
```

Comment:
from previous discussion, I think string and wstring would be structs in C (like sequences) and not just pointers, isnt' it ?

Reply:
@dirk-thomas and @sloretz Did we ever resolve this?

Reply:
Possibly referring to this: https://github.com/ros2/design/pull/190#discussion_r217768639 where array of char and wchar should be structs because they can contain zero?

I don't remember a discussion about wstring or string being structs instead of pointers though.